nano/node/distributed_work.cpp

```diff
@@ -0,0 +1,367 @@
+				if (ec)
+				{
+					this_l->node.logger.try_log (boost::str (boost::format ("Unable to send work_cancel to work_peer %1% %2%: %3% (%4%)") % cancelling_l->address % cancelling_l->port % ec.message () % ec.value ()));
+				}
+			});
+		}
+	});
+}
+
+void nano::distributed_work::stop (bool const local_stop_a)
+{
+	if (!stopped.exchange (true))
+	{
+		std::lock_guard<std::mutex> lock (mutex);
+		if (local_stop_a && (node.config.work_threads != 0 || node.work.opencl))
+		{
+			node.work.cancel (root);
+		}
+		for (auto & connection_w : connections)
+		{
+			if (auto connection_l = connection_w.lock ())
+			{
+				boost::system::error_code ec;
+				connection_l->socket.cancel (ec);
+				if (ec)
+				{
+					node.logger.try_log (boost::str (boost::format ("Error cancelling operation with work_peer %1% %2%: %3%") % connection_l->address % connection_l->port % ec.message () % ec.value ()));
+				}
+				try
+				{
+					connection_l->socket.close ();
+				}
+				catch (const boost::system::system_error & ec)
+				{
+					node.logger.try_log (boost::str (boost::format ("Error closing socket with work_peer %1% %2%: %3%") % connection_l->address % connection_l->port % ec.what () % ec.code ()));
+				}
+			}
+		}
+		connections.clear ();
+		outstanding.clear ();
+	}
+}
+
+void nano::distributed_work::success (std::string const & body_a, boost::asio::ip::address const & address)
+{
+	auto last (remove (address));
+	std::stringstream istream (body_a);
+	try
+	{
+		boost::property_tree::ptree result;
+		boost::property_tree::read_json (istream, result);
+		auto work_text (result.get<std::string> ("work"));
+		uint64_t work;
+		if (!nano::from_string_hex (work_text, work))
+		{
+			uint64_t result_difficulty (0);
+			if (!nano::work_validate (root, work, &result_difficulty) && result_difficulty >= difficulty)
+			{
+				node.unresponsive_work_peers = false;
+				set_once (work);
+				stop (true);
+			}
+			else
+			{
+				node.logger.try_log (boost::str (boost::format ("Incorrect work response from %1% for root %2% with diffuculty %3%: %4%") % address % root.to_string () % nano::to_string_hex (difficulty) % work_text));
+				handle_failure (last);
+			}
+		}
+		else
+		{
+			node.logger.try_log (boost::str (boost::format ("Work response from %1% wasn't a number: %2%") % address % work_text));
+			handle_failure (last);
+		}
+	}
+	catch (...)
+	{
+		node.logger.try_log (boost::str (boost::format ("Work response from %1% wasn't parsable: %2%") % address % body_a));
+		handle_failure (last);
+	}
+}
+
+void nano::distributed_work::set_once (boost::optional<uint64_t> work_a)
+{
+	if (!completed.exchange (true))
+	{
+		callback (work_a);
+	}
+}
+
+void nano::distributed_work::failure (boost::asio::ip::address const & address)
+{
+	auto last (remove (address));
+	handle_failure (last);
+}
+
+void nano::distributed_work::handle_failure (bool const last)
+{
+	if (last)
+	{
+		if (!completed)
```

Comment:
This is from the old code as well, what do you think about doing `if (last && !completed)` instead to improve readability and save an indentation?

Reply:
Agreed, thanks!